builder/azure/arm/step_deploy_template.go

```diff
@@ -108,13 +147,55 @@ func (s *StepDeployTemplate) Cleanup(state multistep.StateBag) {
 		isManagedDisk := state.Get(constants.ArmIsManagedImage).(bool)
 		isSIGImage := state.Get(constants.ArmIsSIGImage).(bool)
 		armStorageAccountName := state.Get(constants.ArmStorageAccountName).(string)
-		imageType, imageName, err := s.disk(ctx, subscriptionId, resourceGroupName, computeName)
+		imageType, imageName, err := s.getDisk(ctx, subscriptionId, resourceGroupName, computeName)
 		if err != nil {
 			ui.Error(fmt.Sprintf("Could not retrieve OS Image details: %s", err))
 		}
-		err = s.delete(ctx, subscriptionId, deploymentName, resourceGroupName)
+
+		deploymentOperations, err := s.listDeploymentOps(ctx, deploymentOpsId)
 		if err != nil {
-			s.reportIfError(err, resourceGroupName)
+			ui.Error(fmt.Sprintf("Could not retrieve deployment operations: %s\n Virtual Machine %s, and its please manually delete it and its ascoiated resources", err, computeName))
+			return
+		}
+		resources := map[string]string{}
+		var vmID *virtualmachines.VirtualMachineId
+		var networkInterfaceID *commonids.NetworkInterfaceId
+		for _, deploymentOperation := range deploymentOperations {
+			// Sometimes an empty operation is added to the list by Azure
+			if deploymentOperation.Properties.TargetResource == nil {
+				continue
+			}
+			resourceName := *deploymentOperation.Properties.TargetResource.ResourceName
+			resourceType := *deploymentOperation.Properties.TargetResource.ResourceType
+
+			// Create vm and NIC ID to use them to delete later
+			switch resourceType {
+			case "Microsoft.Compute/virtualMachines":
+				vmIDDeref := virtualmachines.NewVirtualMachineID(subscriptionId, resourceGroupName, resourceName)
+				vmID = &vmIDDeref
+			case "Microsoft.Network/networkInterfaces":
+				networkInterfaceIDDeref := commonids.NewNetworkInterfaceID(subscriptionId, resourceGroupName, resourceName)
+				networkInterfaceID = &networkInterfaceIDDeref
+			default:
+				resources[resourceType] = resourceName
+			}
+		}
+
+		if vmID != nil {
+			err := s.deleteVM(ctx, *vmID)
+			if err != nil {
+				return
+			}
+		}
+		if networkInterfaceID != nil {
+			err := s.deleteNic(ctx, *networkInterfaceID)
+			if err != nil {
+				return
+			}
+		}
+		err = s.deleteNetworkResources(ctx, subscriptionId, resourceGroupName, resources)
```

Comment:
I see you added a `deleteNetworkResourcesWithQueue` function, was that the one you meant to call here? What's the difference between those two?

Reply:
Also come to think of it, I see we remove the NIC before executing this function, but I would think a NIC is part of the network resources, not saying we should move the NIC deletion to this function necessarily since I believe having it deleted beforehand is the purpose of this PR, but maybe the function should be renamed